### PR TITLE
refactored authz

### LINKED ISSUE #134: Remove dependency on impl in encryption service
**Description**
After the refactor in #132 there is two spots missing where `services` imports stuff from `impl`:
* `enc/storage_handlers.go` 
* `enc/authz_handlers.go`
Both places use the `authz.Authorizer`. The authorizers functionality is kind of mixed in with the `AuthorizeWrapper` and the `AccessObject`. My suggestion would be to see if these functionalities could be consolidated in `impl` somehow. 

**Enabler for**
Following ADR-15

**Acceptance Criteria**
- Dependence on `impl` in `services` is removed. 

**Scope**
Object authorization.

### PR DESCRIPTION
**Pull request by iead-cc**
### Description
After refactoring a small bit o cross contamination was left between services/enc and impl/authz. This PR introduces the neccessary interfaces and disentangles Access Objects and its Authentication

### Parent Issue
Closes #134

### Related Pull Requests
PR #132

### Comments for the Reviewers
- Access Objects and their Authentication has been disentangled
- One Test was removed, the test case is however still covered by "TestAuthorizeWrapperUnauthorized"
- Naming could be improved

### Type(s) of Change (Split the PR if you check many)
- [ ] Added user feature
- [ ] Added technical feature
- [ ] Added tests
- [x] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Infrastructure changes (Terraform, GCP, K8s, other)
- [ ] CI/CD workflow changes

### Testing Checklist
- [x] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [ ] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [x] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make docker-up; make tests`.
- [x] I have updated relevant workflows and deployment tools.
- [x] I have updated/added relevant documentation (readme, doc comments, etc).
- [x] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

File: encryption-service/interfaces/interfaces.go [hunk 1]
```diff
@@ -88,6 +88,35 @@ type UserAuthenticatorInterface interface {
...
+	GetUsers() (userIDs []uuid.UUID, err error)
+
+	// ContainsUser checks if the user is present in the permission list
+	ContainsUser(targetUserID uuid.UUID) (exists bool)
+
+	// getWOEK retrieves the wrapped object encryption key
+	GetWOEK() (woek []byte)
+}
+
+// Interface for authenticating an creating Access Objects
```
Comment (philip-cc) [inline encryption-service/interfaces/interfaces.go:108]:
```suggestion
// Interface for authenticating and creating Access Objects
```

File: encryption-service/interfaces/interfaces.go [hunk 2]
```diff
@@ -88,6 +88,35 @@ type UserAuthenticatorInterface interface {
...
+type AccessObjectAuthenticatorInterface interface {
+	// Creates a new Access Object and inserts it into the Authstorage
+	CreateAccessObject(ctx context.Context, objectID, userID uuid.UUID, woek []byte) (err error)
+
+	// Updates an existing Access Object
```
Comment (philip-cc) [inline encryption-service/interfaces/interfaces.go:113]:
```suggestion
	// Fetches an existing Access Object
```

File: encryption-service/interfaces/interfaces.go [hunk 3]
```diff
@@ -88,6 +88,35 @@ type UserAuthenticatorInterface interface {
...
+	FetchAccessObject(ctx context.Context, objectID uuid.UUID) (accessObject AccessObjectInterface, err error)
+
+	// Updates or inserts the AccessObject into backend storrage
```
Comment (philip-cc) [inline encryption-service/interfaces/interfaces.go:116]:
```suggestion
	// Updates or inserts the AccessObject into backend storage
```

File: encryption-service/impl/authz/access_object.go [hunk 4]
```diff
@@ -71,17 +71,22 @@ func (a *AccessObject) findUser(userID uuid.UUID) (int, bool) {
...
 
-		uids[i] = newUID.String()
+		uids[i] = newUID
 	}
 	return uids, nil
 }
+
+// getWOEK returns the wrapped object encryption key
+func (a *AccessObject) GetWOEK() []byte {
+	return a.Woek
```
Comment (philip-cc) [inline encryption-service/impl/authz/access_object.go:91]: We should make `Woek` private then.

File: encryption-service/services/enc/permissions_handler.go [hunk 5]
```diff
@@ -99,8 +104,9 @@ func (enc *Enc) AddPermission(ctx context.Context, request *AddPermissionRequest
 		return nil, err
 	}
 
+	accessObject.AddUser(target)
 	// Add the permission to the access object
```
Comment (philip-cc) [inline encryption-service/services/enc/permissions_handler.go:108]:
```suggestion
	// Add the permission to the access object
	accessObject.AddUser(target)
```

File: encryption-service/services/enc/permissions_handler.go [hunk 6]
```diff
@@ -148,8 +154,9 @@ func (enc *Enc) RemovePermission(ctx context.Context, request *RemovePermissionR
 		return nil, status.Errorf(codes.InvalidArgument, "invalid target user ID")
 	}
 
+	accessObject.RemoveUser(target)
 	// Add the permission to the access object
```
Comment (philip-cc) [inline encryption-service/services/enc/permissions_handler.go:158]:
```suggestion
	// Add the permission to the access object
	accessObject.RemoveUser(target)
```

File: encryption-service/impl/authz/access_object.go (hunk 4 above)
Comment (iead-cc) [reply encryption-service/impl/authz/access_object.go:91]: we cannot do that as AccessObject is automatically genereated from a proto file. It already has a GetWoek method, so we could reuse that, but I rather have it all capitalized

Comment (philip-cc) [reply encryption-service/impl/authz/access_object.go:91]: Aah, alright!

**Review by philip-cc — APPROVED** [review]

**Pull request merged**